### PR TITLE
HDDS-11380. Fixing the error message of node decommission to be more comprehensive

### PR DESCRIPTION
**Pull request by VarshaRaviCV**
## What changes were proposed in this pull request?
HDDS-11380. Fixing the error message of node decommission to be more comprehensive

Please describe your PR in detail:
Right now, when decommissioning fails quickly due to in-sufficient nodes, the error message says

> Insufficient nodes. Tried to decommission 1 nodes of which 1 nodes were valid. Cluster has 3 IN-SERVICE nodes, 3 of which are required for minimum replication.

This does not clearly explain, how many nodes are required for the decommission to proceed. So in this PR we will be changing the above message to clearly define how many more IN-SERVICE nodes are needed for decommissioning to proceed.

>  Insufficient nodes. Tried to decommission 2 nodes out of 4 IN-SERVICE HEALTHY and 1 UNHEALTHY nodes. Cannot decommission as a minimum of 3 IN-SERVICE HEALTHY nodes are required to maintain replication after decommission.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-11380 

## How was this patch tested?

Ran the existing tests in TestNodeDecommissionManager.java
Also tested locally in docker

sh-4.4$ ozone admin datanode decommission ozone-datanode-4 ozone-datanode-5
Started decommissioning datanode(s):
ozone-datanode-4
ozone-datanode-5
Error: AllHosts: Insufficient nodes. Tried to decommission 2 nodes out of 3 IN-SERVICE HEALTHY and 2 not IN-SERVICE or not HEALTHY nodes. Cannot decommission as a minimum of 3 IN-SERVICE HEALTHY nodes are required to maintain replication after decommission. 
Some nodes could not enter the decommission workflow

sh-4.4$ ozone admin datanode maintenance ozone-datanode-2 ozone-datanode-4 ozone-datanode-5
Entering maintenance mode on datanode(s):
ozone-datanode-2
ozone-datanode-4
ozone-datanode-5
Error: AllHosts: Insufficient nodes. Tried to start maintenance for 3 nodes out of 3 IN-SERVICE HEALTHY and 2 not IN-SERVICE or not HEALTHY nodes. Cannot enter maintenance mode as a minimum of 2 IN-SERVICE HEALTHY nodes are required to maintain replication after maintenance. 
Some nodes could not start the maintenance workflow

### CODE REVIEW

**Review by krishnaasawa1 — COMMENTED** [review]
I believe only if we provide Inservice nodes user is still confused about state of other nodes. e.g: 3 In service but overall 9 . Then what about this 6 nodes why it has not got selected and leads to misinterpretation by user.
Providing information of state of  other node which are not considered will help rule out this confusion.

**Comment by siddhantsangwan** [comment]
Linking our [discussion](https://issues.apache.org/jira/browse/HDDS-11380?focusedCommentId=17879203&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17879203) from the jira here for visibility.

**Comment by VarshaRaviCV** [comment]
@siddhantsangwan @krishnaasawa1 Please review now. I have modified the error message based on review comments and discussions on the jira.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java
Comment (siddhantsangwan) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java:395]: Let's move this computation to the `if` block where the error message is being created. So we perform this computation only if needed.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java
Comment (siddhantsangwan) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java:403]: IntelliJ's suggestion:
```suggestion
          LOG.warn("Cannot decommission {} because it is not IN-SERVICE", dn.getHostName());
```

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java
Comment (siddhantsangwan) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java:408]:
```suggestion
        LOG.warn("Cannot decommission {} because it is not found in SCM", dn.getHostName());
```

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java
Comment (siddhantsangwan) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java:395]: I think what we really need is the total of all nodes that are not (in-service and healthy). That is, `totalNodes - inServiceTotal`. This is because nodes that are entering maintenance, in maintenance, decommissioning etc. also don't count toward in-service healthy nodes.

The easiest way I can see to get `totalNodes` is to use the following API
```
List<DatanodeDetails> getAllNodes();
```
and then get the size.

`NodeStateManager` has a `getTotalNodeCount` API but unfortunately it's not exposed in the interface. We can use `getAllNodes` for now. And later, in another jira, we can expose `getTotalNodeCount` in the interface and use it.

**Review by siddhantsangwan — COMMENTED** [review]
@VarshaRaviCV thanks for working on this. I have added some comments for decommissioning, the same apply for maintenance as well.

**Comment by errose28** [comment]
What is the expected behavior when the max number of nodes are already in decommissioning and maintenance and we try to decom or maintenance mode more nodes? Currently the error message only accounts for in service or unhealthy, but not existing out of service nodes. Also what is the expected message if we try to decom/maintenance a node already in that mode?

Ideally we could add tests for these messages as well. We don't need a full string match, but checking that the message has things like "3 Unhealthy", "2 Healthy", etc for various scenarios would be good.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java
Comment (Tejaskriya) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java:439]: I think we can also show the number of nodes that are not in service, i.e., DECOMMISSIONING, DECOMMISSIONED, ENTERING_MAINTENANCE, IN_MAINTENANCE to show the full picture to the user. Something like this:
```suggestion
                " IN_SERVICE UNHEALTHY and "  + notInService + " decommissioned/in-maintenance nodes. Cannot decommission as a minimum of " + reqNodes +
```

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java
Comment (Tejaskriya) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java:608]: Same suggestions as above

**Review by Tejaskriya — COMMENTED** [review]
There are similar messages for when we try to put a node to maintenance in `checkIfMaintenancePossible()`. We can update them too (in this PR or in another one, either is fine)

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java
Comment (siddhantsangwan) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java:439]: This is all subjective, but I'm cautious about showing a lot of information... it could just end up confusing the user. What do you think about simply showing the number of good nodes (in-service healthy) and the number of bad nodes (not in-service healthy)?

Comment (Tejaskriya) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java:439]: That makes sense.. My thought was basically to show 2 numbers: 
1. the number of datanodes that technically can be decommissioned (without the fail-early logic), so the in-service healthy ones. Which is what you have referred to as good nodes above
2. the number of datanodes that cannot be decommissioned (without the fail-early logic), so the unhealthy and not-in-service nodes

If we don't show the unhealthy number, then the issue for which the jira was raised wouldn't be resolved. Maybe as a part of the bad nodes we can show  (not in-service healthy + in-service unhealthy) number of nodes?

Alternatively, we can have a message like this too:
```
"Insufficient nodes. Tried to decommission " + dns.size() + 
" nodes, of which " + numDecom " are IN_SERVICE. " +  
"Cluster has " + inServiceTotal + " IN-SERVICE HEALTHY and " + unHealthyTotal + " UNHEALTHY nodes. " + 
"Cannot decommission as a minimum of " + reqNodes +
" IN-SERVICE HEALTHY nodes are required in the cluster to maintain replication after decommission. ";
```
What do you think about this?

Comment (siddhantsangwan) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java:439]: Yes, your latest comment is what I meant as well. I agree. One way to get unhealthyTotal is what I commented earlier - https://github.com/apache/ozone/pull/7155#discussion_r1745951180.

**Comment by siddhantsangwan** [comment]
@VarshaRaviCV can we please see the latest error message that's shown on the CLI after your updates?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java
Comment (VarshaRaviCV) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java]: Done.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java
Comment (VarshaRaviCV) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java]: Fixed based on the comments. Now we will show only no of HEALTHY IN-SERVICE and UNHEALTHY (total_nodes-in_service_healthy) nodes.

**Comment by VarshaRaviCV** [comment]
@siddhantsangwan Here are few of the error messages for different scenarios.

sh-4.4$ ozone admin datanode decommission ozone-datanode-1 ozone-datanode-5 ozone-datanode-3
Started decommissioning datanode(s):
ozone-datanode-1
ozone-datanode-5
ozone-datanode-3
Error: AllHosts: Insufficient nodes. Tried to decommission 3 nodes out of 5 IN-SERVICE HEALTHY and 0 UNHEALTHY nodes. Cannot decommission as a minimum of 3 IN-SERVICE HEALTHY nodes are required to maintain replication after decommission. 
Some nodes could not enter the decommission workflow

sh-4.4$ ozone admin datanode decommission ozone-datanode-4 ozone-datanode-5
Started decommissioning datanode(s):
ozone-datanode-4
ozone-datanode-5
Error: AllHosts: Insufficient nodes. Tried to decommission 2 nodes out of 3 IN-SERVICE HEALTHY and 2 UNHEALTHY nodes. Cannot decommission as a minimum of 3 IN-SERVICE HEALTHY nodes are required to maintain replication after decommission. 
Some nodes could not enter the decommission workflow

sh-4.4$ ozone admin datanode maintenance ozone-datanode-2 ozone-datanode-4 ozone-datanode-5
Entering maintenance mode on datanode(s):
ozone-datanode-2
ozone-datanode-4
ozone-datanode-5
Error: AllHosts: Insufficient nodes. Tried to start maintenance for 3 nodes out of 3 IN-SERVICE HEALTHY and 2 UNHEALTHY nodes. Cannot enter maintenance mode as a minimum of 2 IN-SERVICE HEALTHY nodes are required to maintain replication after maintenance. 
Some nodes could not start the maintenance workflow

**Review by Tejaskriya — APPROVED** [review]
LGTM!

**Review by siddhantsangwan — APPROVED** [review]
LGTM. @VarshaRaviCV thanks for improving this! Pending green CI.

**Comment by errose28** [comment]
From [this comment](https://github.com/apache/ozone/pull/7155#issuecomment-2332329106), can you share the output for decommissioning an already decommissioning or in maintenance node and decommissioning a nonexistent node?

We should also not merge this without unit tests on the numbers in the error message. From what I see current tests in `TestNodeDecommissionManager` only cover success/failure of different scenarios.

**Comment by siddhantsangwan** [comment]
I'm expecting these messages to show up if the datanode is in maintenance or missing.
```
 } catch (NodeNotFoundException e) {
        // We already validated the host strings and retrieved the DnDetails
        // object from the node manager. Therefore we should never get a
        // NodeNotFoundException here expect if the node is remove in the
        // very short window between validation and starting decom. Therefore
        // log a warning and ignore the exception
        LOG.warn("The host {} was not found in SCM. Ignoring the request to " +
            "decommission it", dn.getHostName());
        errors.add(new DatanodeAdminError(dn.getHostName(),
            "The host was not found in SCM"));
      } catch (InvalidNodeStateException e) {
        errors.add(new DatanodeAdminError(dn.getHostName(), e.getMessage()));
      }
    }
    return errors;
  }
```

**Comment by VarshaRaviCV** [comment]
@siddhantsangwan @Tejaskriya @errose28 I have added assertions for node count in error message in the existing tests of TestNodeDecommissionManager.java. Please check once.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java
Comment (adoroszlai) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java:401]: I think this message (and similar one for maintenance) could be improved by including the actual `opState` the node is in.

Also, please consider replacing `dn.getHostName()` with `dn` to rely on:

https://github.com/apache/ozone/blob/f784a84a04bbabde01f3154d24a1eb92344088d1/hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/protocol/DatanodeDetails.java#L561-L563

This shows just a bit more info to make the node uniquely identifiable while keeping the hostname for convenience.  (Applies to all similar logs.)

**Review by adoroszlai — COMMENTED** [review]
Thanks @VarshaRaviCV for the patch.

**Comment by adoroszlai** [comment]
@VarshaRaviCV do you plan to update the patch based on my [suggestion](https://github.com/apache/ozone/pull/7155#discussion_r1804310374), or should we do it in a follow-up?

**Comment by VarshaRaviCV** [comment]
@adoroszlai I will pick up the [suggestion](https://github.com/apache/ozone/pull/7155#discussion_r1804310374) in a follow-up PR since this PR already has approvals.

**Comment by adoroszlai** [comment]
Thanks @VarshaRaviCV for the patch, @errose28, @krishnaasawa1, @siddhantsangwan, @Tejaskriya for the review.

**Pull request merged**
